src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -191,224 +204,233 @@ private void CreateStreamAndStartWatcher()
-                // and use it for pattern matching. We'll go back to the eventPaths[i] to get the actual path for notifications
-                string path = eventPaths[i];
-                if (path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase))
+                Debug.Assert((numEvents.ToInt32() == eventPaths.Length) && (numEvents.ToInt32() == eventFlags.Length) && (numEvents.ToInt32() == eventIds.Length));
+
+                // Try to get the actual watcher from our weak reference.  We maintain a weak reference most of the time
+                // so as to avoid a rooted cycle that would prevent our processing loop from ever ending
+                // if the watcher is dropped by the user without being disposed. If we can't get the watcher,
+                // there's nothing more to do (we can't raise events), so bail.
+                FileSystemWatcher watcher;
+                if (!_weakWatcher.TryGetTarget(out watcher))
                 {
-                    path += "/.";
+                    CancellationCallback();
                 }
 
-                // First, we should check if this event should kick off a re-scan since we can't really rely on anything after this point if that is true
-                if (ShouldRescanOccur(eventFlags[i]))
-                {
-                    OnError(new ErrorEventArgs(new IOException(SR.FSW_BufferOverflow, (int)eventFlags[i])));
-                    break;
-                }
-                else if ((handledRenameEvents != null) && (handledRenameEvents.Contains(eventIds[i])))
-                {
-                    // If this event is the second in a rename pair then skip it
-                    continue;
-                }
-                else if ((DoesPathPassNameFilter(path)) && ((_filterFlags & eventFlags[i]) != 0))
+                // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
+                // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
+                List<FSEventStreamEventId> handledRenameEvents = null;
+
+                for (long i = 0; i < numEvents.ToInt32(); i++)
                 {
-                    // The base FileSystemWatcher does a match check against the relative path before combining with 
-                    // the root dir; however, null is special cased to signify the root dir, so check if we should use that.
-                    string relativePath = null;
-                    if (eventPaths[i].Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase) == false)
+                    // Match Windows and don't notify us about changes to the root folder itself
+                    string path = eventPaths[i];
+                    if (path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase))
```

Comment:
Fixing this issue for both Linux and OS X would be a little more work than I expected...I'll revert the part of the change around the active issue and have a separate PR for it